Hw2 task4 - cache function implementation

IuliiaLanchuk opened the pull request.

Cjkjvfnby commented on homework2/task04/project4/task4.py:
```diff
@@ -0,0 +1,49 @@
+    return (a ** b) ** 2
+
+
+def multiply(a: int, b: int) -> int:
```
Please check `multiply` translation

Cjkjvfnby commented on homework2/task04/project4/task4.py:
```diff
@@ -0,0 +1,49 @@
+from typing import Callable
+
+
+def function(a: int, b: int) -> int:
```
You can move this and other function declarations to the tests.

Cjkjvfnby commented on homework2/task04/project4/task4.py:
```diff
@@ -0,0 +1,49 @@
+
+def multiply(a: int, b: int) -> int:
+    """Return result of arguments sum."""
+    return a + b
```
Actually, you don't need to declare your own functions, you can use any existing.  For example `operator.add` or `sum`

Cjkjvfnby commented on homework2/task04/tests/test_cach.py:
```diff
@@ -0,0 +1,27 @@
+
+cache_function = cache(function)
+cache_multiply = cache(multiply)
+cache_operations = cache(operations)
```
This function is actually pretty the same.    Lest use something different, for example sum` and [1, 2].

Cjkjvfnby commented on homework2/task04/tests/test_cach.py:
```diff
@@ -0,0 +1,27 @@
+cache_operations = cache(operations)
+
+some = 2, 4
+val_1 = cache_function(*some)
```
Inline these variables.  It is better to define them inside the test.

Cjkjvfnby commented on homework2/task04/tests/test_cach.py:
```diff
@@ -0,0 +1,27 @@
+    assert val_2 is val_5
+
+
+def test_sum():
```
Please check that the same function with the different arguments is not cached. 

Something like
```python
cache_function(1, 2) is not cache_function (2, 1) 
```

Cjkjvfnby replied:
The idea was to test that when you call the function with different arguments, you don't reuse cached results.

Cjkjvfnby commented on homework2/task04/project4/task4.py:
```diff
@@ -0,0 +1,34 @@
+    """Return function."""
+    cache_data = {}
+
+    def wrapper(*args: int) -> int:
```
why int?

IuliiaLanchuk replied:
Yes, you are right. It can be also string, list and so on.

Cjkjvfnby commented on homework2/task04/tests/test_cach.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def test_sum():
+    def summa(*args) -> int:
```
You don't need to write your own `summ`, just cache `sum`

IuliiaLanchuk replied:
Ok, fixed

Cjkjvfnby commented on homework2/task04/tests/test_cach.py:
```diff
@@ -0,0 +1,43 @@
+)
+def test_cache(func: Callable, data: Tuple[int], expected_result: bool):
+    temp_result = cache(func)
+    assert (temp_result(*data) is temp_result(*data)) == expected_result
```
`is` does not work as you expect, try to use Mock to test if a function was called.  

```
func = Mock()
cache(func)
...
```

IuliiaLanchuk replied:
Done

Cjkjvfnby commented on homework2/task04/tests/test_cach.py:
```diff
@@ -0,0 +1,53 @@
+)
+def test_cache_same_values(func: Callable):
+    some = [2, 4, 7]
+    mock_func = Mock(func)
```
It is a Mock, you don't need function.

IuliiaLanchuk replied:
ok, fixed

Cjkjvfnby commented on homework2/task04/tests/test_cach.py:
```diff
@@ -0,0 +1,53 @@
+    mock_func = Mock(func)
+    cache_calc_sum_of_values = cache(mock_func)
+
+    cache_calc_sum_of_values(*some)
```
What is * mean here?

IuliiaLanchuk replied:
Here (*some) transforms list [a, b] into separate a, b values, in this turn
(*args) gives all positional arguments as a tuple.  Tuples are immutable and hashable and able to use with dictionary. List is unhashable type, so it can't be used with `if args not in cache_data`.

Cjkjvfnby commented:
I have requested a specific input in one of the comments. I still did not see it in tests.

Cjkjvfnby commented:
What about this request?
https://github.com/IuliiaLanchuk/hometasksEpam/pull/10#discussion_r589469118

Cjkjvfnby commented on homework2/task04/tests/test_cach.py:
```diff
@@ -0,0 +1,42 @@
+def test_cache_sum():
+    some = [1, 2]
+    cache_calc_sum_of_values = cache(sum)
+    assert cache_calc_sum_of_values(*some) == cache_calc_sum_of_values(*some)
```
When I ask to use `[1, 2]` as an argument, I mean that `[1, 2]` should be passed as an argument, not unpacked to a function as a tuple of arguments.

Cjkjvfnby commented on homework2/task04/project4/task4.py:
```diff
@@ -0,0 +1,35 @@
+
+    def wrapper(*args: Any) -> Any:
+        """Return cache data."""
+        str_args = str(args)
```
This approach may be used but can you give me at least two different examples of when this function will fail?

IuliiaLanchuk commented:
Have you seen my fixes? My last commit

Cjkjvfnby reviewed (changes requested):
Sorry, I forget to press submit button. 

Code it ok, but I still have a question.

IuliiaLanchuk commented:
https://github.com/IuliiaLanchuk/hometasksEpam/pull/10#discussion_r596941099
@Cjkjvfnby 
1) The main idea of this approach is to get func arguments + result and store them in cache as a key-value pair. In this approach we store key as string value using `str()` convertion, so we need to be sure that converted object implements `__str__` method otherwise str() call will return object reference which not reflect function input. 

2) Moreover, we need to be sure that `str` will have idempotent behavior, means will return same key for any number of calls with same arguments. In this turn,` __str__` function should be defined correctly in order to return unique value for each different input.

Cjkjvfnby commented:
> so we need to be sure that converted object implements __str__ method otherwise str() call will return object reference which not reflect function input ...

I agree with that conditions.  Could you give me an example of 2 Python objects (builtin) that implement str in the way you described, but they will lead to a wrong results?

IuliiaLanchuk reviewed (commented):
> > so we need to be sure that converted object implements **str** method otherwise str() call will return object reference which not reflect function input ...
> 
> I agree with that conditions. Could you give me an example of 2 Python objects (builtin) that implement str in the way you described, but they will lead to a wrong results?
@Cjkjvfnby 
for example, object(), zip().

Cjkjvfnby commented:
> List is unhashable type, so it can't be used with if args not in cache_data.

That was a point why I asked you to pass a list to that function :)

Cjkjvfnby commented:
Let's got back to my question.
```
def foo(x):
    return x

cached(foo)(x) -> x
cached(foo)(y) -> x
```
Please provide an example of x  and y.

IuliiaLanchuk commented:
> Let's got back to my question.
> 
> ```
> def foo(x):
>     return x
> 
> cached(foo)(x) -> x
> cached(foo)(y) -> x
> ```
> 
> Please provide an example of x and y.
@Cjkjvfnby 
to follow this example, I implemented class TestClass with single argument `number` and default method `__str__` always returning same string value '1'. In following test I created two differnet objects TestClass with number as 10 and 12. So, I expect that cache will have two different records with two objects respectively, but since I define `__str__` method as mentioned above I will have only one cached record.
Actually, to reach this scenario I did one more update in my code. I changed convertion to string from `str(args)` to `str(*args)`. I did it because `args` is a tuple object that has his own `__str__` realisation which will not give us such effect.

![image](https://user-images.githubusercontent.com/64098113/113345265-e425c980-933a-11eb-8ccc-d475642c5f0f.png)

![image](https://user-images.githubusercontent.com/64098113/113345305-f1db4f00-933a-11eb-88a1-8653254492b5.png)

`

Cjkjvfnby commented:
> to follow this example, 

- Does it fit your requirements for __str__ implementation?  
- You don't need to write your own class, to fail this code it is enough to use existing. 
- Have you test how `str(*args)` works with 2 arguments?

Cjkjvfnby commented:
Please insert code with code tag, not as screenshots

![image](https://user-images.githubusercontent.com/171597/113357520-ccefd780-934c-11eb-9f9d-7abde0a672c3.png)

IuliiaLanchuk commented:
https://github.com/IuliiaLanchuk/hometasksEpam/pull/10#issuecomment-812189056
@Cjkjvfnby 
1) Of course this example does not fit `__str__` requirements due to always returning the same string object.
This example was to show the failing scenario.
2) **added two extra tests `test_same_cache_diff_args_filter`, `test_same_cache_diff_args_zip`**
3) It will fail because str() can take only 1 argument

`def test_same_cache_diff_args_filter():`
`    def foo(x):`
`        return x`
`                                                 `
`    cache_function = cache(foo)`
`    l1 = ['4', '4', 'd']`
`    f_first = **filter**(str.isdigit, l1)`
`    f_second = f_first`
`    assert list(f_first) != list(f_second)`    #  list(f_first)=['4', '4'],  list(f_second)=[]
`    assert cache_function(f_first) == cache_function(f_second)`


`def test_same_cache_diff_args_zip():`
`    def foo(x):`
`        return x`
`                                              `
`    cache_function = cache(foo)`
`    l1 = ['4', '4', 'd']`
`    z_first = **zip**(l1)`
`    z_second = z_first`
 `   assert list(z_first) != list(z_second)`
 `   assert cache_function(z_first) == cache_function(z_second)`

Cjkjvfnby commented:
> added two extra tests 

You have not pushed it yet.

https://github.com/IuliiaLanchuk/hometasksEpam/pull/10#issuecomment-811369250
I expect a very simple response. It is not something complicated. 

```python
x = ?
y = ? 
```

IuliiaLanchuk commented:
Maybe you mean 1 and '1' ? the same result str() but diff args?

Cjkjvfnby commented:
> Maybe you mean 1 and '1' ? the same result str() but diff args?

Exactly.
